File: bkend/book.py
import os
import shutil
import zipfile

from bs4 import BeautifulSoup


class Book:
    def __init__(self, pth, des, by_zip=False):
        self.pth = pth
        self.des = des
        self.by_zip = by_zip
        self.Cover = None
        self.Name = None
        self.Chapters = None

    def sayhell(self):
        print("HELLL NOOO")

    def init(self):
        fileName = os.path.basename(self.pth)
        isFile = os.path.exists(self.pth)

        print(fileName, isFile)
        if fileName.endswith(".epub") and isFile:
            k = fileName.split(".")
            k.pop()
            print(k)
            m = ".".join(k)
            k = [m]

            # Create a temporary ZIP file
            temp_zip_path = os.path.join(self.des, k[0] + ".zip")
            shutil.copyfile(self.pth, temp_zip_path)

            # Extract the EPUB contents
            with zipfile.ZipFile(temp_zip_path, "r") as zip_ref:
                zip_ref.extractall(os.path.join(self.des, k[0]))

            os.remove(temp_zip_path)

            return 1
        else:
            print("Not an EPUB file")
            return 0

    def get_cover(self, fld=None):
        fileName = os.path.basename(self.pth)
        if fld:
            k = [fld]
        else:
            k = fileName.split(".")
            k.pop()
            m = ".".join(k)
            k = [m]

        book_folder = os.path.join(self.des, k[0])
        flds = os.listdir(book_folder)
        flds = [f for f in flds if not f.startswith(".")]  # Remove hidden files

        opf_present = [f for f in flds if f.endswith(".opf")]
        print(opf_present, "here ")
        for f in flds:
            know = os.path.isdir(os.path.join(self.des, k[0], f))
            inside_files = []
            if know:
                inside_files = os.listdir(os.path.join(book_folder, f))
                inside_files = [f for f in inside_files if not f.startswith(".")]
                print([f for f in inside_files if f.endswith(".opf")])

            if (
                not opf_present
                and any(f.endswith(".opf") for f in inside_files)
                and know
            ):
                opf_present = False
                book_files = os.listdir(os.path.join(self.des, k[0], f))

                for i in book_files:
                    if i.endswith(".opf"):
                        print("in the opf")
                        opf_file_path = os.path.join(self.des, k[0], f, i)
                        with open(opf_file_path, "r", encoding="utf-8") as opf_file:
                            soup = BeautifulSoup(opf_file, "xml")
                            meta_cover = soup.find("meta", attrs={"name": "cover"})
                            book_name = soup.find("dc:title").text.strip()

                            print(meta_cover)

                            self.Name = book_name
                            print(book_name, meta_cover)
                            if meta_cover and meta_cover["name"] == "cover":
                                cover_id = meta_cover["content"]
                                pic = soup.find("item", attrs={"id": cover_id})
                                if pic:
                                    url = pic.get("href")
                                    real_link = os.path.join(self.des, k[0], f, url)
                                    self.Cover = real_link.replace("\\", "/")
                                    return url
                            else:
                                self.Cover = "NotFound"
            else:
                pass

        if opf_present and len(opf_present) > 0:
            book_files = os.listdir(os.path.join(self.des, k[0]))
            for i in book_files:
                if i.endswith(".opf"):
                    opf_file_path = os.path.join(self.des, k[0], i)
                    with open(opf_file_path, "r", encoding="utf-8") as opf_file:
                        soup = BeautifulSoup(opf_file, "xml")
                        meta_cover = soup.find("meta", attrs={"name": "cover"})
                        book_name = soup.find("dc:title").text.strip()

                        if not meta_cover:
                            meta_cover = soup.find("opf:meta", attrs={"name": "cover"})

                        self.Name = book_name
                        if meta_cover and meta_cover["name"] == "cover":
                            cover_id = meta_cover["content"]
                            pic = soup.find("item", attrs={"id": cover_id})
                            if pic:
                                url = pic.get("href")
                                real_link = os.path.join(self.des, k[0], url)
                                self.Cover = real_link.replace("\\", "/")
                                return url
                        else:
                            self.Cover = "NotFound"

    def book_data(self, fld=None):
        fileName = os.path.basename(self.pth)
        if fld:
            k = [fld]
        else:
            k = fileName.split(".")
            k.pop()
            m = ".".join(k)
            k = [m]

        if_exists = os.path.exists(os.path.join(self.des, k[0]))

        if if_exists:
            print(k)

            book_folder = os.path.join(self.des, k[0])
            flds = os.listdir(book_folder)
            print(flds)
            flds = [f for f in flds if not f.startswith(".")]
            op_present = [f for f in flds if f.endswith(".ncx")]
            print(op_present, "here")

            for f in flds:
                know = os.path.isdir(os.path.join(self.des, k[0], f))
                print(know, "know")

                inside_files = []
                if know:
                    inside_files = os.listdir(os.path.join(book_folder, f))
                    inside_files = [f for f in inside_files if not f.startswith(".")]
                    print([f for f in inside_files if f.endswith(".ncx")], inside_files)

                if (
                    len(op_present) == 0
                    and any(f.endswith(".ncx") for f in inside_files)
                    and know
                ):
                    op_present = False
                    book_files = os.listdir(os.path.join(self.des, k[0], f))

                    for i in book_files:
                        if i.endswith(".ncx"):
                            ncx_file_path = os.path.join(self.des, k[0], f, i)
                            with open(ncx_file_path, "r", encoding="utf-8") as ncx_file:
                                soup = BeautifulSoup(ncx_file, "xml")
                                nav_content = soup.select("navMap navPoint")

                                lessons = []
                                for point in nav_content:
                                    name = (
                                        point.find("navLabel")
                                        .text.strip()
                                        .replace("  ", "")
                                        .replace("\t", "")
                                        .replace("\n", "")
                                    )
                                    link = point.find("content")["src"]
                                    real_link = os.path.join(self.des, k[0], f, link)
                                    lessons.append(
                                        {
                                            "name": name,
                                            "link": real_link.replace("\\", "/"),
                                        }
                                    )

                                self.Chapters = lessons
                                return {
                                    "Chapters": lessons,
                                }
                else:
                    print("Something wrong")

            if op_present and len(op_present) > 0:
                book_files = os.listdir(os.path.join(self.des, k[0]))
                for i in book_files:
                    if i.endswith(".ncx"):
                        ncx_file_path = os.path.join(self.des, k[0], i)
                        with open(ncx_file_path, "r", encoding="utf-8") as ncx_file:
                            soup = BeautifulSoup(ncx_file, "xml")
                            nav_content = soup.select("navMap navPoint")

                            lessons = []
                            for point in nav_content:
                                name = (
                                    point.find("navLabel")
                                    .text.strip()
                                    .replace("  ", "")
                                    .replace("\t", "")
                                    .replace("\n", "")
                                )
                                link = point.find("content")["src"]
                                real_link = os.path.join(self.des, k[0], link)
                                lessons.append(
                                    {
                                        "name": name,
                                        "link": real_link.replace("\\", "/"),
                                    }
                                )

                            self.Chapters = lessons
                            return {
                                "Chapters": lessons,
                            }
        else:
            print("Book folder does not exist")

    def saythings(self):
        print(self.pth, self.des)


# Example usage:
# bk = Book(
#     "E:\\books\\e.epub",
#     "C:\\Users\\vikle\\Documents\\GitHubProjects\\Epub-book-Reader\\bkend\\books",
# )
# # await bk.init()
# p = bk.get_cover("me")
# k = bk.book_data("me")
# print(k)
# print(p)
# bk.book_data()

# Modify paths and call methods as needed

File: node js backend/book.js
const path = require("path");
const fs = require("fs");

const cheerio = require("cheerio");

const extract = require("extract-zip");

class Book {
  pth;
  des;
  Book;
  byZip;
  Cover;
  Name;
  Chapters;

  constructor(p, d, Zip = false) {
    this.pth = p;
    this.des = d;
    console.log(Zip);
    if (Zip) {
      this.byZip = true;
    } else {
      this.byZip = false;
    }
  }

  async sayhell() {
    console.log("HELLL NOOO  ");
  }

  async init() {
    let fileName = path.basename(this.pth);
    let isFile = await fs.existsSync(this.pth);

    console.log(fileName, isFile);
    if (fileName.includes(".epub") && isFile) {
      let k = fileName.split(".");
      k.pop();
      console.log(k);
      let m = k.join(".");
      k = [m];
      let copyfile = await fs.copyFileSync(
        this.pth,
        path.join(this.des, k[0]) + ".zip"
      );

      await fs.rmSync(path.join(this.des, k[0]), {
        recursive: true,
        force: true,
      });
      await fs.mkdirSync(path.join(this.des, k[0]));

      await extract(path.join(this.des, k[0]) + ".zip", {
        dir: path.join(this.des, k[0]),
      });

      // try {
      //   await fs
      //     .createReadStream(path.join(this.des , k[0]) + ".zip")
      //     .pipe(unzipper.Extract({ path: path.join(this.des , k[0]) }))
      //     .promise();
      //   console.log("Extraction complete");
      // } catch (error) {
      //   console.error("Error during extraction:", error);
      // }
      await fs.rmSync(path.join(this.des, k[0]) + ".zip", { force: true });
      return 1;
    } else {
      console.log("not an epubfile");
      return 0;
    }
  }

  async getCover(fld) {
    let fileName = path.basename(this.pth);
    let k;
    if (fld !== undefined) {
      k = [fld];
    } else {
      k = fileName.split(".");
      k.pop();
      let m = k.join(".");
      k = [m];
    }

    // console.log(flds , 'this here is were ')
    let bookFolder = path.join(this.des, k[0]);
    let flds = await fs.readdirSync(bookFolder);
    flds = flds.filter((e) => e !== ".DS_Store");
    // flds.pop();

    let opPresent = flds.filter(isOpf);

    function isOpf(file) {
      return path.basename(file).includes(".opf");
    }

    for (let f of flds) {
      let know = await fs.lstatSync(path.join(this.des, k[0], f)).isDirectory();

      let insideFiles = [];
      if (know) {
        insideFiles = await fs.readdirSync(path.join(bookFolder, f));
        insideFiles = insideFiles.filter((e) => e !== ".DS_Store");
        console.log(insideFiles.filter(isOpf));
      }

      if (
        opPresent.length === 0 &&
        insideFiles.filter(isOpf).length > 0 &&
        know
      ) {
        console.log("in the ops");
        opPresent = false;
        let book = await fs.readdirSync(path.join(this.des, k[0], f));

        for (let i of book) {
          if (i.includes(".opf")) {
            console.log(path.join(this.des, k[0], f, i));
            let rd = await fs.readFileSync(
              path.join(this.des, k[0], f, i),
              "utf8"
            );
            const $ = cheerio.load(rd);

            let navContent = $('meta[name~="cover"]');
            let bookName = $("metadata").find("dc\\:title").text();

            console.log(bookName.length, navContent.length);
            if (navContent.length === 0) {
              console.log("IN HERE");
              navContent = $('opf\\:meta[name~="cover"]');
            }

            if (bookName.length === 0) {
              bookName = $("opf\\:metadata").find("dc\\:title").text();
            }

            this.Name = bookName;
            console.log(bookName + "bkbkb");

            console.log(navContent.attr("name"));

            let nm = navContent.attr("name");
            let cv = navContent.attr("content");

            console.log(cv + "here cv");

            if (nm === "cover") {
              let pic = $(`item[id~="${cv}"]`);

              let url = pic.attr("href");

              console.log(url, " mew");
              let realLink = path.join(this.des, k[0], f) + "\\" + url;
              this.Cover = realLink.replaceAll(/\\/g, "/");
              return url;
            } else if (nm === undefined) {
              let cov = $("#cover").attr("href");
              console.log(cov, "here2");

              if (cov !== undefined) {
                let realLink = path.join(this.des, k[0], f, cov);
                this.Cover = realLink.replaceAll(/\\/g, "/");
              } else {
                this.Cover = "NotFound";
              }
            } else {
              this.Cover = "NotFound";
            }
          }
        }
      } else {
        console.log("DAM ");
      }
    }

    if (opPresent.length > 0) {
      let book = await fs.readdirSync(path.join(this.des, k[0]));

      for (let i of book) {
        if (i.includes(".opf")) {
          let rd = await fs.readFileSync(
            path.join(this.des, k[0]) + "\\" + i,
            "utf8"
          );
          const $ = cheerio.load(rd);
          const navContent = $('meta[name~="cover"]');
          const bookName = $("metadata").find("dc\\:title").text();

          console.log(bookName.length, navContent.length);
          if (navContent.length === 0) {
            console.log("IN HERE");
            navContent = $('opf\\:meta[name~="cover"]');
          }

          if (bookName.length === 0) {
            bookName = $("opf\\:metadata").find("dc\\:title").text();
          }

          this.Name = bookName;

          console.log(navContent.attr("name"), "lol");

          console.log(bookName + "bkbkb");

          console.log(navContent.attr("name"));

          let nm = navContent.attr("name");
          let cv = navContent.attr("content");

          console.log(cv + "here cv");

          if (nm === "cover") {
            let pic = $(`item[id~="${cv}"]`);

            let url = pic.attr("href");

            console.log(url, " mew");
            let realLink = path.join(this.des, k[0]) + "\\" + url;
            this.Cover = realLink.replaceAll(/\\/g, "/");
            return url;
          } else if (nm === undefined) {
            let cov = $("#cover").attr("href");
            if (cov !== undefined) {
              let realLink = path.join(this.des, k[0], cov);
              this.Cover = realLink.replaceAll(/\\/g, "/");
            } else {
              this.Cover = "NotFound";
            }
          } else {
            this.Cover = "NotFound";
          }
        }
      }
    }
  }
  async bookData(fld) {
    let fileName = path.basename(this.pth);
    let k;
    if (fld !== undefined) {
      k = [fld];
    } else {
      k = fileName.split(".");
      k.pop();
      let m = k.join(".");
      k = [m];
    }
    let ifExists = await fs.existsSync(path.join(this.des, k[0]));

    if (ifExists) {
      console.log(k);

      let bookFolder = path.join(this.des, k[0]);
      let flds = await fs.readdirSync(bookFolder);
      console.log(flds);
      flds = flds.filter((e) => e !== ".DS_Store");
      let opPresent = flds.filter(isNcx);
      console.log(opPresent, "here");
      function isNcx(file) {
        return file.includes(".ncx");
      }

      //   flds.pop(".DS_Store");
      console.log(flds, "k");
      for (let f of flds) {
        let know = await fs
          .lstatSync(path.join(this.des, k[0], f))
          .isDirectory();
        console.log(know, "know");

        let insideFiles = [];
        if (know) {
          insideFiles = await fs.readdirSync(path.join(bookFolder, f));
          insideFiles.pop(".DS_Store");

          console.log(insideFiles.filter(isNcx), "m", insideFiles);
        }

        if (
          opPresent.length === 0 &&
          insideFiles.filter(isNcx).length > 0 &&
          know
        ) {
          opPresent = false;
          let book = await fs.readdirSync(path.join(this.des, k[0], f));

          for (let i of book) {
            if (i.includes(".ncx")) {
              let rd = await fs.readFileSync(
                path.join(this.des, k[0], f, i),
                "utf8"
              );
              const $ = cheerio.load(rd);
              const navContent = $("navMap navPoint");

              let lessons = [];
              for (let p of navContent) {
                let point = $(p);

                let name = point.find("navLabel").text().replaceAll("  ", "");
                name = name.replaceAll(/\t/g, "");
                name = name.replaceAll(/\n/g, "");
                let link = point.find("content").attr("src");
                let realLink = path.join(this.des, k[0], f) + "\\" + link;
                lessons.push({
                  name: name,
                  link: realLink.replaceAll(/\\/g, "/"),
                });
              }

              this.Chapters = lessons;
              return {
                Chapters: lessons,
              };
            }
          }
        } else {
          console.log("something wrong ");
        }
      }
      console.log(opPresent.length, opPresent, "le");
      if (opPresent.length > 0) {
        let book = await fs.readdirSync(path.join(this.des, k[0]));

        for (let i of book) {
          if (i.includes(".ncx")) {
            let fil = path.join(this.des, k[0], i);
            let rd = await fs.readFileSync(fil, "utf8");
            const $ = cheerio.load(rd);
            const navContent = $("navMap navPoint");

            let lessons = [];
            for (let p of navContent) {
              let point = $(p);

              let name = point.find("navLabel").text().replaceAll("  ", "");
              name = name.replaceAll(/\t/g, "");
              name = name.replaceAll(/\n/g, "");
              let link = point.find("content").attr("src");
              let realLink = path.join(this.des, k[0]) + "\\" + link;

              lessons.push({
                name: name,
                link: realLink.replaceAll(/\\/g, "/"),
              });
            }

            this.Chapters = lessons;
            return {
              Chapters: lessons,
            };
          }
        }
      }
    }
  }

  saythings() {
    console.log(this.pth, this.des);
  }
}

// let bk = new Book(
//   "/Users/pablo/Downloads",
//   "/Users/pablo/Downloads/Epub-book-Reader-EPubJS-Integration/books"
// );

// let bk = new Book(
//   "C:\\Users\\vikle\\Documents\\GitHubProjects\\Epub-book-Reader\\books",
//   "C:\\Users\\vikle\\Documents\\GitHubProjects\\Epub-book-Reader\\books"
// );
// // // // // await bk.init();
// // // Matt Parker - Humble Pi_ A Comedy of Maths Errors-Penguin UK (2019)
// async function somethign() {
//   let n = await bk.getCover(
//     "Matt Parker - Humble Pi_ A Comedy of Maths Errors-Penguin UK (2019)"
//   );
//   await bk.bookData(
//     "Matt Parker - Humble Pi_ A Comedy of Maths Errors-Penguin UK (2019)"
//   );
//   console.log(bk.Name, bk.Cover, " here");
// }

// somethign();

module.exports = Book;
